### ISSUE TITLE
Async Closure not FnMut anymore after nightly-2024-02-06

### ISSUE THREAD

**Issue by aWeinzierl** [rust-lang/rust#120841]
I recently updated my Rust binaries using rustup, and found that it breaks at compilation unexpectedly on nightly-2024-02-07 and later.
I created the following reduced code which still replicates the compile errors.

### Code

I tried this code:

```rust
#![feature(async_closure)]

use std::future::Future;

pub trait FutureFactory<T> {
    fn new(&mut self) -> T;
}

impl<T, TT> FutureFactory<TT> for T
where
    T: Unpin + FnMut() -> TT,
{
    #[allow(clippy::new_ret_no_self)]
    fn new(&mut self) -> TT {
        (self)()
    }
}

pub fn new<T>(_factory: impl FutureFactory<T>) -> () { }

async fn _do_stuff<Resolver, F>(
    resolve: Resolver,
) -> ()
where
    Resolver: Fn() -> F + Copy + Send + 'static,
    F: Future<Output = ()> + Send + 'static,
{
    let _result = new(async || resolve());
}
fn main() {}

```

I expected to see this happen: compilation success

Instead, this happened: compilation error:
```
error[E0277]: expected a `FnMut()` closure, found `{coroutine-closure@src/main.rs:28:23: 28:31}`
  --> src/main.rs:28:23
   |
28 |     let _result = new(async || resolve());
   |                   --- ^^^^^^^^^^^^^^^^^^ expected an `FnMut()` closure, found `{coroutine-closure@src/main.rs:28:23: 28:31}`
   |                   |
   |                   required by a bound introduced by this call
   |
   = help: the trait `FnMut<()>` is not implemented for `{coroutine-closure@src/main.rs:28:23: 28:31}`, which is required by `{coroutine-closure@src/main.rs:28:23: 28:31}: FutureFactory<_>`
   = note: wrap the `{coroutine-closure@src/main.rs:28:23: 28:31}` in a closure with no arguments: `|| { /* code */ }`
note: required for `{coroutine-closure@src/main.rs:28:23: 28:31}` to implement `FutureFactory<_>`
  --> src/main.rs:9:13
   |
9  | impl<T, TT> FutureFactory<TT> for T
   |             ^^^^^^^^^^^^^^^^^     ^
10 | where
11 |     T: Unpin + FnMut() -> TT,
   |                ------------- unsatisfied trait bound introduced here
note: required by a bound in `new`
  --> src/main.rs:19:30
   |
19 | pub fn new<T>(_factory: impl FutureFactory<T>) -> () { }
   |                              ^^^^^^^^^^^^^^^^ required by this bound in `new`

For more information about this error, try `rustc --explain E0277`.
```

### Version it worked on

It most recently worked on: channel `nightly-2024-02-06-x86_64-pc-windows-msvc`

### Version with regression

channel `nightly-2024-02-07-x86_64-pc-windows-msvc` or later (including current nightly of Rust playground)

`rustc --version --verbose`:
```
rustc 1.78.0-nightly (256b6fb19 2024-02-06)
binary: rustc
commit-hash: 256b6fb19a2c018eaad4806d2369d1f6a71fc6ec
commit-date: 2024-02-06
host: x86_64-pc-windows-msvc
release: 1.78.0-nightly
LLVM version: 17.0.6
```

**Comment by Nilstrieb** [comment]
@compiler-errors this is probably your doing. I don't know whether this is expected or not.

**Comment by compiler-errors** [comment]
@aWeinzierl: Yes, this is expected; however, it should be fixed by #120712
